sw/device/lib/testing/kmac/kmac_vector_generate.c

```diff
@@ -0,0 +1,67 @@
+// Copyright lowRISC contributors.
```

Comment:
Question: does this test vector generator need to be its own stand alone device program? How is it intended to be run?

Currently it is in `sw/device/` which would imply it is is designed to run on an opentitan device. However, it is a `cc_binary` target, which indicates it is a host binary.

If this is a standalone host tool then perhaps it should go under `sw/host/`. If this is suppose to be an on-device utility, then perhaps the `main` and helper functions should become testutils under `sw/device/lib/testing/kmac_testutils` where they can be used by tests that are run via the OTTF. Does that make sense?

Reply:
Having `kmac_vector_generate` as a host program makes more sense (therefore I will be moving it under `sw/host`). I just added this as an example use of a library, and this program is not really doing anything useful for the moment. I think it makes more sense to improve it a bit so that it becomes a sanity check on vendoring the library, as @dmcardle suggested below (if I didn't misunderstand that comment).

Eventually what I am hoping is to automate test generation by XKCP side and feed it to KMAC (from cryptolib, #16936). Since I do not know how CI works, I think I am not there yet (not sure if CI is the suitable place to test it either, in case it might break). Since KMAC supports 4 algorithms with varying-size inputs, having some randomly generated inputs at every run would make more sense for testing. We already have some [hard-coded tests](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/drivers/kmac_test_vectors.c) but their coverage is fairly limited.

Reply:
I see, so eventually you would potentially expand the tests here to use this XKCP library to generate random test vectors on the fly (presumably with some sort of seed so the test failures are reproducible)?

If that is the case, then perhaps there is no need for this host testing binary (`kmac_vector_generate`) att all? Could you can just have this PR vendor in the code (with Bazel as you have done) and then have a follow up PR use this library in a `opentitan_functest` that is an on-device test? You as long as the test does not have a  `long` or `eternal` timeout, it should run on the FPGA in CI.

Reply:
I removed the host binary `kmac_vector_generate` like you suggested. I agree that test generating can be better handled in a follow-up PR. This PR now only brings the XKCP bazel scripts.

Generating vectors directly on-device from a known seed could also work. Do you happen to know if it is feasible to generate long test vectors (e.g. something on the order of ~50 KB) directly on the device? I had just assumed that generating such long vectors on the device would not be feasible because of its runtime or memory budget. Although, come to think of it, generating such a large vector on the host and passing it to the device is also another challenge.

Reply:
I guess we should try and see. I'd prefer generating these on the device for the sake of simplicity.

Reply:
Sounds good, I will try to see if we can generate KMAC vectors on the device in reasonable time. Thanks for suggestions @timothytrippel  and @alphan.